File: leetcode/queue_stack/3141_Maximum_Hamming_Distances/codes/queue.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> maxHammingDistances(vector<int>& nums, int m) {
        int n = 1 << m;
        vector<int> dist(n, -1);
        queue<int> q;
        for (int x : nums) {
            dist[x] = 0;
            q.push(x);
        }
        while (!q.empty()) {
            int u = q.front(); q.pop();
            for (int b = 0; b < m; b++) {
                int v = u ^ (1 << b);
                if (dist[v] == -1) {
                    dist[v] = dist[u] + 1;
                    q.push(v);
                }
            }
        }
        int mask = n - 1;
        vector<int> ans;
        ans.reserve(nums.size());
        for (int x : nums) {
            ans.push_back(m - dist[mask ^ x]);
        }
        return ans;
    }
};

File: leetcode/queue_stack/3141_Maximum_Hamming_Distances/codes/queue.py
from collections import deque
from typing import List

class Solution:
    def maxHammingDistances(self, nums: List[int], m: int) -> List[int]:
        n = 1 << m
        dist = [-1] * n
        q = deque()
        for x in nums:
            dist[x] = 0
            q.append(x)
        while q:
            u = q.popleft()
            for b in range(m):
                v = u ^ (1 << b)
                if dist[v] == -1:
                    dist[v] = dist[u] + 1
                    q.append(v)
        mask = n - 1
        ans = []
        for x in nums:
            ans.append(m - dist[mask ^ x])
        return ans
